### ISSUE TITLE
How do I use the `route` function in a compiled Vue component?

### ISSUE THREAD

**Issue by simensen** [tightenco/ziggy#70]
I'm using the following code:

```
                <div class="list-group" v-if="projects.length > 0">
                    <a v-for="project in projects" :href="route('api.project', {project_id: project.id})" class="list-group-item">{{ project.name }}</a>
                </div>
```

But I get the following error in console:

![image](https://user-images.githubusercontent.com/191200/30065751-83cb7176-921b-11e7-880a-6cef790256f2.png)

Note that both `route` and `namedRoutes` exist. I have `@route` above the `app.js` line in my layout.

**Comment by stevebauman** [comment]
I experience this issue right now.

You have to use computed properties to utilize the `route()` method in your components.

It seems Vue inits faster than the route helper or something? I'm not a very experienced JS dev so I'm not sure why it doesn't register when it's above my main app script.

I hope this is changed in the future so we can compile it with the rest of our components.

**Comment by simensen** [comment]
@stevebauman Great, thanks. I decided to use a function instead as that appears to work OK:

```
                <div class="list-group" v-if="projects.length > 0">
                    <a v-for="project in projects" :href="projectUrl(project)" class="list-group-item">{{ project.name }}</a>
                </div>
```

...

```
<script>
    export default {
        methods: {
            projectUrl(project) {
                return route('api.project', {project_id: project.id})
            },
...
```

Indeed, it would be nice to be able to use this the other way, though. I can keep this issue open as a feature request or I can close it. I'm happy that I have a solution now.

**Comment by ankurk91** [comment]
It is vue.js issues, when you directly pass global method in prop, vue thinks that this is local method and throw error when not found within component.

**Comment by DanielCoulbourne** [comment]
Yeah I imagine passing window.route() would work as well. Maybe we should add a config option to import the module in the traditional es6 way

**Comment by stevebauman** [comment]
Definitely interested in ES6 module :)

**Comment by olimorris** [comment]
@DanielCoulbourne FYI I came to use Ziggy for the specific purpose of what @simensen had posted. I modified slightly:

```
methods: {
    projectUrl(named_route, options) {
        return window.route(named_route, options)
    }
}
```

so I can do:

```
<a :href="projectUrl('plan_subscribe', { plan: plan.id })">Subscribe</a>
```

More than happy to create a pull request to your readme. I'm not a VueJS expert so I have no idea whether the above is considered good practice.

**Comment by DanielCoulbourne** [comment]
@olimorris Any reason not to just do 
```
<a :href="window.route('plan_subscribe', { plan: plan.id })">Subscribe</a>
```
And skip the intermediate step?

**Comment by olimorris** [comment]
@DanielCoulbourne with Vue 2.4.4 it throws this error:

```
Property or method "window" is not defined on the instance but referenced during render. Make sure to declare reactive data properties in the data option.
```

From that I deduced that I needed to add an intermediate step and hence the method in my Vue object. I guess you could always include a seperate Vue component for people to use within Vue which references `route`.

**Comment by DanielCoulbourne** [comment]
Gotcha. Thanks! I'm mostly a react guy so it didn't occur to me. I'll consider making an includable Vue component. If you have an example of how you'd want it implemented I'd love to see what you're thinking.

**Comment by stevebauman** [comment]
@DanielCoulbourne If you create an ES6 module then we could simply attach the module to our global Vue instance using:

> **Note**: The below is just an example, but we can't bind $route / $router to the Vue instance since those are reserved, so I chose `$api` as the property.

```js
import route from 'ziggy'

Vue.prototype.$api = route;
```

In our components:

```js
export default {
    methods: {
        create() {
            let url = this.$api('api.posts.store');
            
            this.$http.post(url).then(() => {
                // Created
            });
        }
    }
}
```

**Comment by DanielCoulbourne** [comment]
@stevebauman Would you want to pair and talk though this sometime?

**Comment by Archer70** [comment]
I know this issue is old, but I came across it by Googling the same problem and thought I might add my solution as well. I was able to make a global `vueRoute` function by adding this to app.js

```javascript
Vue.mixin({
    methods: {
        vueRoute: route
    }
});
```

And then you can call it like so:
```html
<a class="nav-link" :href="vueRoute('home')">{{ $t('Home') }}</a>
```

**Comment by DanielCoulbourne** [comment]
Great solution! I'm going to reopen this issue and add this to the docs next time I do a sweep

**Comment by montogeek** [comment]
@DanielCoulbourne How are you using it in a React app?
I ran `php artisan ziggy:generate`, iimported the created file, that is just an object, How could I import `route`?

**Comment by montogeek** [comment]
I could use `@routes` directives, then use it from global `window` object (`window.route`), but I don't want to pollute my HTML with that huge, non minified webpack bundle script tag :/

**Comment by DanielCoulbourne** [comment]
Right. you should be able to
```js
import route from 'path/to/ziggy'
```
or just use `Ziggy.

**Comment by montogeek** [comment]
@DanielCoulbourne Default export is a simple static object

**Comment by montogeek** [comment]
Or do you mean `vendor/ziggy/path/to/ziggy`?

**Comment by montogeek** [comment]
This is the static object that I am talking about: https://github.com/tightenco/ziggy/blob/master/src/BladeRouteGenerator.php#L40

**Comment by DanielCoulbourne** [comment]
right. I see.

In your app.js `import Ziggy from 'path/to/exported/js'`

or if you want to be more explicit: `window.Ziggy = 'path/to/exported/js'`

**Comment by DanielCoulbourne** [comment]
If you have to, for some reason, you can also pass a custom route object to ziggy, as in these tests:

https://github.com/tightenco/ziggy/blob/master/tests/js/test.route.js#L621

**Comment by prasetyo-jeanu** [comment]
add route method on mixin
@montogeek 

`
import Vue from 'vue';
import {Ziggy} from './ziggy';
import route from '../../../vendor/tightenco/ziggy/src/js/route';

Vue.mixin({
    methods: {
        route: function ( name, params, absolute ) { return route(name, params, absolute, Ziggy); }
    }
});
`
https://github.com/tightenco/ziggy/issues/154#issuecomment-397841738

**Comment by jny986** [comment]
> add route method on mixin
> @montogeek
> 
> `
> import Vue from 'vue';
> import {Ziggy} from './ziggy';
> import route from '../../../vendor/tightenco/ziggy/src/js/route';
> 
> Vue.mixin({
> methods: {
> route: function ( name, params, absolute ) { return route(name, params, absolute, Ziggy); }
> }
> });
> `
> [#154 (comment)](https://github.com/tightenco/ziggy/issues/154#issuecomment-397841738)

slight fix to this one.

`import route from '../../../vendor/tightenco/ziggy/src/js/route';`
should be
`import route from '../../vendor/tightenco/ziggy/src/js/route';`

**Comment by RUDIWER** [comment]
Can't get it to work with vue3 in laravel ? anybody a idea?

**Comment by bakerkretzmar** [comment]
@RUDIWER feel free to open a new issue with a detailed description of the problem you're having and we'll be happy to help!

**Comment by fabriciomoeller** [comment]
@RUDIWER 

I have the same problem in Vue 3. Did you open a ticket or find the solution ?

**Comment by hackur** [comment]
I was able to get this working with Vue 3 using this issue thread as a reference.

In app.js I added the following:
```
app.mixin({
    methods: {
        route
    }
})
```

Which is the Vue 3 equivalent of `Vue.mixin()` which was used in the examples above. 

I did not have to install the ziggy-js library to get access to it, and this fixed the issue where the `route` method was not available in compiled components.
